### ISSUE TITLE
ECS  – zero downtime problem

### ISSUE THREAD

**Issue by neugeeug** [TeliaSoneraNorge/telia-terraform-modules#36]
In case when we (NEO) do some changes in regards to ECS EC2 instances e.g. changing AMI or type of EC2 instances (Terraform), the cluster behaves as depicted in attached picture. The new ones EC2 instances are in initializing state but the old ones that should still serve resources, are in shutting down state. The result of that is that we have about minute service unavailable. We wish EC2 instances would be changed without any downtime.
![image](https://user-images.githubusercontent.com/32057389/40043256-4490cc20-5824-11e8-9bfc-1049b8522db9.png)

**Comment by neugeeug** [comment]
comment from @itsdalmo 

> That is true. The ECS module uses our ASG module under the hood, which does do rolling updates using Cloudformation. The trick to zero downtime is to only signal CF once your new resources are ready to serve traffic - however, if memory serves, this is not implemented yet for ECS. Will have to look at the module again when I'm in front of my computer, but I think that it waits for the ECS agent to be operational, which of course is not the same as ECS having scheduled any containers on the new instances.

**Comment by itsdalmo** [comment]
Commented a lot on Slack before remembering that you made an issue for this - here is the rundown:

- Kenneth linked this: [AWS Blog](https://aws.amazon.com/blogs/compute/how-to-automate-container-instance-draining-in-amazon-ecs/).
- Anton has a module for that [here](https://github.com/terraform-community-modules/tf_aws_ecs_instance_draining_on_scale_in)
- I linked [Lifecycled](https://github.com/buildkite/lifecycled), which I'm using for Concourse.

I'm thinking blog post (and the module linked by @antonbabenko) is the best way to solve this, since the Lambda talks directly to ECS to figure out whether any containers are running on the instance being terminated. Since Cloudformation will not continue with the rolling update until instances targeted for termination have been properly terminated, you should have zero downtime if the Lambda does it's job.

**Comment by itsdalmo** [comment]
Started looking at this now, but I noticed that the module lives on community modules and not aws modules - the prior being in the process of getting deprecated?

I need to change [this line](https://github.com/terraform-community-modules/tf_aws_ecs_instance_draining_on_scale_in/blob/master/main.tf#L115) in the module so that we can have more than one such function in any account. Made a fork of the repo when I saw that it was on the old community modules, and I'm wondering if it should be moved and/or further cleaned up before we use it?

Thoughts @antonbabenko ?

**Comment by antonbabenko** [comment]
👍 terraform-community-modules is still a managed location for some modules, but it is ok to use your fork for some time. I need to do some cleanup and publish it in terraform-aws-modules. Probably tomorrow.

**Comment by itsdalmo** [comment]
I've deployed the ecs/example using modules from PR https://github.com/TeliaSoneraNorge/telia-terraform-modules/pull/40, and ran some quick tests by manually tweaking the desired count on the autoscaling group that was created. Some quick notes/takeaways:

- The Lambda (python code) is very chatty. It's hard to follow the logs.
- It's recursive and has roughly 1 invocation per second from receiving the lifecycle hook event until it completes. Since each invocation produces ~50 log lines, it's hard to find any useful info (and debug log level is hardcoded).
- In my trials, it took ~8 minutes to scale in from 3 to 2 instances, which only required stopping 1 container (which was serving no traffic).

I was hoping it would be quicker to move the containers. Not sure what's going on here, but it would basically be quicker to do this manually 😝 If the 8minutes is "fixed" and not a fluke, it could take a long time to do rolling updates like this...

**Comment by itsdalmo** [comment]
I'm thinking the pullrequest I linked above needs to be tested "in the wild". You up for that @neugeeug?

**Comment by neugeeug** [comment]
with pleasure :) I will let you know
